sdk/src/main/scala/io/horizen/account/state/ForgerStakeStorage.scala

```diff
@@ -0,0 +1,358 @@
+package io.horizen.account.state
+
+import com.google.common.primitives.Bytes
+import io.horizen.account.proposition.AddressProposition
+import io.horizen.account.state.ForgerStakeLinkedList.{addNewNode, getStakeListItem, linkedListNodeRefIsNull, removeNode}
+import io.horizen.account.state.ForgerStakeStorage.saveStorageVersion
+import io.horizen.account.state.ForgerStakeStorageVersion.ForgerStakeStorageVersion
+import io.horizen.account.state.NativeSmartContractMsgProcessor.NULL_HEX_STRING_32
+import io.horizen.account.state.WithdrawalMsgProcessor.calculateKey
+import io.horizen.account.utils.BigIntegerUtil
+import io.horizen.account.utils.WellKnownAddresses.FORGER_STAKE_SMART_CONTRACT_ADDRESS
+import io.horizen.evm.Address
+import io.horizen.proposition.{PublicKey25519Proposition, VrfPublicKey}
+import sparkz.crypto.hash.Blake2b256
+
+import java.math.BigInteger
+import scala.util.{Failure, Success}
+
+trait ForgerStakeStorage {
+
+  def existsStakeData(view: BaseAccountStateView, stakeId: Array[Byte]): Boolean = {
+    // do the RAW-strategy read even if the record is actually multi-line in stateDb. It will save some gas.
+    val data = view.getAccountStorage(FORGER_STAKE_SMART_CONTRACT_ADDRESS, stakeId)
+    // getting a not existing key from state DB using RAW strategy
+    // gives an array of 32 bytes filled with 0, while using CHUNK strategy
+    // gives an empty array instead
+    !data.sameElements(NULL_HEX_STRING_32)
+  }
+
+  def findStakeData(view: BaseAccountStateView, stakeId: Array[Byte]): Option[ForgerStakeData] = {
```

Comment:
Maybe we could move this into the object companion below. In this way we can avoid getting the specific versions of storage class in the message processor

Reply:
I see your point, but I would prefer to keep them in the trait, in case in the future we had another storage version with a different way to store the stake data. This way, these methods can be easily overridden, without the need to change the client code.